### PR TITLE
Issue #3683 - ensure multipart files are still cleaned up if error occurs

### PR DESCRIPTION
**Pull request by lachlan-roberts**
Issue #3683

- separated the parsing from the constructor of `MultiParts` so the `Request.__MULTIPARTS` attribute can be set before a parsing error occurs

- added test to replicate the issue where the files are not deleted when an error occurs

- general cleanups of `MultiParts` and `MultiPartFormInputStream`

### CODE REVIEW

File: jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java [hunk 1]
```diff
@@ -330,18 +328,38 @@ public String getContentDispositionFilename()
...
+
+        if (contextTmpDir == null)
+            contextTmpDir = new File(System.getProperty("java.io.tmpdir"));
+
         if (_config == null)
-            _config = new MultipartConfigElement(_contextTmpDir.getAbsolutePath());
-        
+            _config = __DEFAULT_MULTIPART_CONFIG;
+
+        if (!"".equals(_config.getLocation()))
```
Comment (olamy) [inline jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java]: we now have ` StringUtil.isEmpty ` :)

File: jetty-http/src/test/java/org/eclipse/jetty/http/MultiPartFormInputStreamTest.java [hunk 2]
```diff
@@ -69,8 +69,7 @@ public MultiPartFormInputStreamTest()
     }
     
     @Test
-    public void testBadMultiPartRequest()
-            throws Exception
```
Comment (janbartel) [inline jetty-http/src/test/java/org/eclipse/jetty/http/MultiPartFormInputStreamTest.java]: There are no actual changes to this file, other than formatting changes. This clutters up the PR and makes it hard to follow what changes were actually necessary for the bug fix, so can we leave formatting changes to be done only where there is a substantive change to the file.

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java [hunk 3]
```diff
@@ -747,11 +747,14 @@ protected void initMultiPart () throws Exception
...
         {
+            if (LOG.isDebugEnabled())
+                LOG.debug("multipart cleanup listener added for {}", this);
+
             //Register a listener to delete tmp files that are created as a result of this
             //servlet calling Request.getPart() or Request.getParts()
-
             ContextHandler ch = ContextHandler.getContextHandler(getServletHandler().getServletContext());
-            ch.addEventListener(MultiPartCleanerListener.INSTANCE);
+            if(!Arrays.asList(ch.getEventListeners()).contains(MultiPartCleanerListener.INSTANCE))
```
Comment (janbartel) [inline jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java]: ServletHolder.initServlet() should only be called once - do you have a repro test case where it is called multiple times?

File: jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java [hunk 4]
```diff
@@ -372,57 +390,29 @@ public boolean isEmpty()
         return true;
     }
     
-    /**
-     * Get the already parsed parts.
-     *
-     * @return the parts that were parsed
-     */
-    @Deprecated
-    public Collection<Part> getParsedParts()
```
[the remark this answers is not in the document]
Comment (lachlan-roberts) [reply jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java]: `MultiPartFormInputStream` was a newly created class to replace the old `MultiPartInputStreamParser` so I don't think we needed to import the deprecated methods from the old class in the first place. But I will add it back.

File: jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java [hunk 5]
```diff
@@ -123,8 +122,8 @@ protected void close() throws IOException
         {
             _out.close();
         }
-        
-        protected void write(int b) throws IOException
+
+        protected void write(byte b) throws IOException
```
[the remark this answers is not in the document]
Comment (lachlan-roberts) [reply jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java]: yes, this is just part of the cleanups I was doing for the `MultiPartFormInputStream`, this method is not used but I think it is correct for it to be taking a byte and not an int

[the remark this answers is not in the document]
Comment (sbordet) [reply jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java]: @lachlan-roberts please revert this change, as all the stream APIs (`OutputStream`, etc.) to write a single byte take an `int`, not a `byte`.
Furthermore, this single-byte write must be implemented to forward the call to `write(byte[], int, int)` to avoid code duplication.

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java (hunk 3 above)
Comment (lachlan-roberts) [reply jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java]: I don't currently have a test for this, Greg seemed to think that this could be called for different servlets using the same ContextHandler.

**Comment by lachlan-roberts** [comment]
@janbartel do you want me to remove all changes not directly related to fixing the issue and propose the cleanup changes in a separate PR

File: jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (janbartel) [reply jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java]: OK, if its a new class, then it shouldn't have any deprecated methods :)

**Comment by janbartel** [comment]
@lachlan-roberts bingo!

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java (hunk 3 above)
Comment (janbartel) [reply jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java]: You (and Greg) are perfectly correct - without this test multiple servlets that support multipart could add the instance multiple times.

**Comment by lachlan-roberts** [comment]
the `MultiPartServletTest` written for this does not work on windows, have not been able to work out why
```
Caused by: 
java.io.IOException: Could Not Delete File
	at org.eclipse.jetty.http.MultiPartFormInputStream$MultiPart.delete(MultiPartFormInputStream.java:292)
	at org.eclipse.jetty.http.MultiPartFormInputStream$MultiPart.cleanUp(MultiPartFormInputStream.java:303)
	at org.eclipse.jetty.http.MultiPartFormInputStream.deleteParts(MultiPartFormInputStream.java:407)
	at org.eclipse.jetty.server.MultiParts$MultiPartsHttpParser.close(MultiParts.java:77)
	at org.eclipse.jetty.server.MultiPartCleanerListener.requestDestroyed(MultiPartCleanerListener.java:51)
```

**Comment by janbartel** [comment]
Are you sure that the file is closed before you try to delete it? Windows is sensitive about these things.

**Comment by olamy** [comment]
@lachlan-roberts I can see ```         public InputStream getInputStream() throws IOException
        {
            if (_file != null)
            {
                // written to a file, whether temporary or not
                return new BufferedInputStream(new FileInputStream(_file));
            }
            else
            {
                // part content is in memory
                return new ByteArrayInputStream(_bout.getBuf(), 0, _bout.size());
            }
        }  ```
is this InputStream closed?

**Comment by lachlan-roberts** [comment]
The tests are now passing on windows as well, the part was not being closed if an exception was happening during parsing
